diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 011e868..c9c7c84 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -2440,7 +2442,6 @@ static void build_ntdll_module(void)
     node_ntdll = wm->ldr.DdagNode;
     if (TRACE_ON(relay)) RELAY_SetupDLL( module );
 
-    hidden_exports_init( wm->ldr.FullDllName.Buffer );
 }
 
 
@@ -2991,6 +2992,7 @@ static WINE_MODREF *build_main_module(void)
     if (status) goto failed;
     RtlFreeUnicodeString( &nt_name );
     wm->ldr.LoadCount = -1;
+    hidden_exports_init( wm->ldr.FullDllName.Buffer );
     return wm;
 failed:
     MESSAGE( "wine: failed to create main module for %s, status %lx\n",

